### PR TITLE
Jokes

### PR DESCRIPTION
**Pull request by echase6**

### CODE REVIEW

File: practice/jokes/jokes/models.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        return self.setup == other.setup and self.punchline == other.punchline
+
+    def __repr__(self):
+        r""" Magic repr function.
+
+        >>> Joke('knock-knock', 'who dat?')
+        Joke(knock-knock, who dat?)
+        """
+        return 'Joke({}, {})'.format(self.setup, self.punchline)
```
Comment (davidselassie) [inline practice/jokes/jokes/models.py]: Minor point: I encourage reprs to be valid Python code. If you use `Joke({!r}, {!r}'.format(self.setup, self.punchline)` you'll get the correct quotes around the strings so you can instantiate that Joke.

File: practice/jokes/jokes/models.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def get_all_jokes():
+    r"""Return all of the jokes.
+
+    >>> get_all_jokes()
+    [['knock-knock', 'who 2 dat?'], ['bahbah', 'blacksheep']]
+    """
+    return [
+        [joke.setup, joke.punchline]
```
Comment (davidselassie) [inline practice/jokes/jokes/models.py]: What is the purpose of unpacking the Joke class instances? In the rest of your code you can use them, including the templates. It's okay for a model class to be used in the template.

File: practice/jokes/jokes/models.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    if _are_both_parts_entered(setup, punchline):
+        new_joke = Joke(setup, punchline)
+        _JOKES.append(new_joke)
+        return new_joke
+    else:
+        raise ValueError('joke or punchline is invalid')
+
+
+def _are_both_parts_entered(setup, punchline):
```
Comment (davidselassie) [inline practice/jokes/jokes/models.py:76]: Minor point: think about naming your functions for their qualitative purpose. `_is_joke_valid()` explains what the function is doing, but actually leaves the details to the implementation.

File: practice/jokes/jokes/static/jokes/site.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+var setups = $('.setup');
+
+/**
+ * Add click handlers to each of the setups
+ */
+ var click_handler = setups.on('click', function() {
+   return $(this).next().show()
+ });
+
+setups.append(click_handler)
```
Comment (davidselassie) [inline practice/jokes/jokes/static/jokes/site.js]: I don't think this last line does anything. Calling `.append()` on an element section adds a new inner element. I don't know what it does when you add a list of click handlers?

File: practice/jokes/jokes/models.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    Traceback (most recent call last):
+    ...
+    ValueError: joke or punchline is invalid
+    """
+    if _are_both_parts_entered(setup, punchline):
+        new_joke = Joke(setup, punchline)
+        _JOKES.append(new_joke)
+        return new_joke
+    else:
+        raise ValueError('joke or punchline is invalid')
```
Comment (davidselassie) [inline practice/jokes/jokes/models.py:73]: Good work providing a full set of named interface functions for manipulating jokes to the rest of the program. Consider using the [functional core, state-ful shell](https://github.com/PdxCodeGuild/Full-Stack-Day-Class/blob/master/sessions/2016-07-07/tips/2016-08-31.md#functional-core-stateful-shell) trick to make this function more testable.

File: practice/jokes/jokes/models.py (hunk 2 above)
Comment (echase6) [reply practice/jokes/jokes/models.py]: Good point.  I like that change.

**Pull request merged**